### PR TITLE
Add support for `Expect-CT` HTTP header

### LINKED ISSUE #321: Support for Expect-CT header
This new header informs the browser that it is expected for your certificate to appear in a Certificate Transparency log.

Detailed explanation here: https://scotthelme.co.uk/a-new-security-header-expect-ct/
Browser support: proposed in Chrome, planned for Firefox (https://www.chromestatus.com/feature/5677171733430272)
Specification: https://datatracker.ietf.org/doc/draft-stark-expect-ct/

Sample header: `Expect-CT: enforce; max-age=30; report-uri https://scotthelme.report-uri.io/r/default/ct/enforce`

If `enforce` then the site will not be loaded if the certificate cant be found in a CT log. If absent, then it will only be reported.

### PR DESCRIPTION
**Pull request by jacobbednarz**
`Expect-CT` is a new security header that performs a request to a
certificate transparency log endpoint that will check whether the
presented certificate adhears to the UA's certificate transparency
policy.

Whilst building this out, I made the decision to use the hash
configuration syntax (similar to CSP) instead of the string based syntax
(like HSTS) to ensure we could validate and build various parts of the
value without needing to perform horrible regexes across it.

Fixes #321

### CODE REVIEW

**Comment by jacobbednarz** [comment]
I've got this running in our CI suite at the moment and it's all 💚  here

File: lib/secure_headers.rb [hunk 1]
```diff
@@ -51,6 +52,7 @@ def opt_out?
   CSP = ContentSecurityPolicy
 
   ALL_HEADER_CLASSES = [
+    ExpectCt,
```
Comment (oreoshake) [inline lib/secure_headers.rb]: I prefer to spell out the full name of the header spec.

File: lib/secure_headers/configuration.rb [hunk 2]
```diff
@@ -116,7 +116,7 @@ def deep_copy_if_hash(value)
 
     attr_writer :hsts, :x_frame_options, :x_content_type_options,
       :x_xss_protection, :x_download_options, :x_permitted_cross_domain_policies,
-      :referrer_policy, :clear_site_data
+      :referrer_policy, :clear_site_data, :expect_ct
```
Comment (oreoshake) [inline lib/secure_headers/configuration.rb]: I prefer to spell out the full name of the header spec.

File: lib/secure_headers/headers/expect_ct.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
+module SecureHeaders
+  class ExpectCtConfigError < StandardError; end
+
+  class ExpectCt
+    HEADER_NAME = 'Expect-CT'.freeze
+    CONFIG_KEY  = :expect_ct
+    INVALID_CONFIGURATION_ERROR = 'config must be a hash.'.freeze
```
Comment (oreoshake) [inline lib/secure_headers/headers/expect_ct.rb]: I prefer `"`

File: lib/secure_headers/headers/expect_ct.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        enforced_directive,
+        max_age_directive,
+        report_uri_directive
+      ].compact.join('; ').strip
+    end
+
+    def enforced_directive
+      # Unfortunately `if @enforced` isn't enough here in case someone
+      # passes in a random string so let's be specific with it to prevent
+      # accidental enforcement.
```
Comment (oreoshake) [inline lib/secure_headers/headers/expect_ct.rb]: :+1:

File: lib/secure_headers/headers/expect_ct.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      "enforce" if @enforced == true
+    end
+
+    def max_age_directive
+      "max-age=#{@max_age}" if @max_age
+    end
+
+    def report_uri_directive
+      "report-uri=\"#{@report_uri}\"" if @report_uri
```
Comment (oreoshake) [inline lib/secure_headers/headers/expect_ct.rb]: `report-uri` is quoted. Interesting.

File: lib/secure_headers.rb (hunk 1 above)
Comment (jacobbednarz) [inline lib/secure_headers.rb]: Good call! Fixed via 39cbb52

File: lib/secure_headers/headers/expect_ct.rb (hunk 5 above)
Comment (jacobbednarz) [inline lib/secure_headers/headers/expect_ct.rb]: This did strike me as odd too but, you know, specs? `¯\_(ツ)_/¯`

File: lib/secure_headers/headers/expect_ct.rb (hunk 3 above)
Comment (jacobbednarz) [inline lib/secure_headers/headers/expect_ct.rb]: Oh my! What caveman would ever commit files with mixed quotes?!?! 😛  Fixed in 9faf299

**Comment by oreoshake** [comment]
whoops, can you add an assertion [here](https://github.com/twitter/secureheaders/blob/a85b80cca27106f357900f0ac930caec30f6be74/spec/spec_helper.rb#L36) that says this header is not applied by default.

**Comment by jacobbednarz** [comment]
Of course! Added in c265600.

**Comment by jacobbednarz** [comment]
We've been running this for over a week now and all is looking good on our front 👍

**Comment by jacobbednarz** [comment]
This is pending release from one of the browser vendors to test the functionality confirms to the outlined spec. My comment above was merely, "this isn't blowing up our application with memory leaks or such".

**Comment by jacobbednarz** [comment]
updates! https://groups.google.com/a/chromium.org/forum/m/#!msg/blink-dev/7_EgKuWAXjE/QJCzbBscBQAJ

**Comment by jacobbednarz** [comment]
Looks like this is coming in Chrome 61
https://blog.chromium.org/2017/08/chrome-61-beta-javascript-modules.html<https://blog.chromium.org/2017/08/chrome-61-beta-javascript-modules.html?m=1>
________________________________

**Comment by oreoshake** [comment]
Released in 3.7.0

**Pull request merged**